### ISSUE TITLE
Refactor where function

### ISSUE THREAD

**Issue by johngrimes** [aehrc/pathling#116]
Reimplement the where function as part of the refactored design of the v3 implementation.

Tasks:

- [x] Make join helpers support joining of grouped datasets
- [x] Join from distinct identities to argument dataset filtered of null values
- [x] Investigate why there seem to be multiple stages loading the same Parquet file, with the same filters

**Comment by johngrimes** [comment]
An additional test case, courtesy of @ametke:

```
reverseResolve(Observation.subject).where($this.code.coding contains http://purl.obolibrary.org/obo/hp.owl|HP:0001250 and $this.valueCodeableConcept.coding contains http://snomed.info/sct|52101004).count() > 0
```

**Comment by johngrimes** [comment]
Closed by #140.
